3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,202 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {map} from '../src/utils/object';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {reportError} from '../src/error';
+initLogConstructor();
+setReportError(reportError);
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * This provides the "glue" between the AMP Analytics tag and the third-party
+ * vendor's metrics-collection page.
+ *
+ * The vendor's page must implement onNewAmpAnalyticsInstance().
+ */
+class AmpAnalytics3pRemoteFrameHelper {
+  constructor(win) {
+    /**
+     * @type {!Array<function(string)>}
+     * @private
+     */
+    this.eventsListeners_ = [];
+
+    /**
+     * @type {!Array<function(string,string)>}
+     * @private
+     */
+    this.extraDataListeners_ = [];
+
+    /**
+     * @type {!Array<Object>}
+     * @private
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @type {!Array<Object>}
+     * @private
+     */
+    this.receivedExtraDataQueue_ = [];
+
+    /**
+     * Holds a mapping between sender ID and extra data
+     * @type {Map}
+     * @private
+     */
+    this.senderIdToExtraData_ = map();
+
+    /**
+     * Simple requestIdleCallback polyfill
+     * @type {Function<Function>}
+     * @private
+     */
+    this.requestIdleCallback_ =
+      win.requestIdleCallback.bind(win) || (cb => setTimeout(cb, 1));
+
+    /**
+     * Handles communication between frames
+     * @type {IframeMessagingClient}
+     * @private
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(JSON.parse(window.name).sentinel);
+    this.iframeMessagingClient_.registerCallback('ampAnalytics3pMessages',
+      this.onReceivedMessagesFromCreative_.bind(this));
+
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!Function} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventsListeners_.push(listener);
+  }
+
+  /**
+   * Registers a callback function to be called when extra setup data is
+   * recevied from an AMP Analytics tag.
+   * @param {!Function} listener A function to receive any extra
```

Comment:
Actually @param {!function(!Array)} since yesterday's commits. But the change you suggest above will remove this method anyway...